[Accessibility] Add required a11y text and decoration to required fields

amesplant opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #48948 .
- Update forms that the `wc-template-functions` and Address changes to ensure that the asterisk is seen but not read by a screen reader (`aria-required="true"` will announce that the field is required) 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Open `/my-account/edit-account/` and using a screen reader, tab to each field. Make sure that the screen reader reads each fields label, required (if required), along with any descriptive text below the fields.
2. Edit each of the addresses `/my-account/edit-address/billing/` and `/my-account/edit-address/shipping/`, and follow the same process as you did in step 1, making sure that the screen reader correctly reads each field as you tab through the form.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

amesplant-dmv committed:
Add required a11y text and decoration to required fields

jorgeatorres commented:
@amesplant: Thanks for your help with the a11y issues. Is this specific PR ready for review? If so, could you please include some testing instructions? There are also a few PHP linting issues that would need to be addressed, as well as some conflicts with `trunk`. Let me know :)

amesplant commented:
@jorgeatorres not quite ready .. in discussion in the issue #48948

amesplant committed:
Merge remote-tracking branch 'upstream/trunk' into fix/48948-required-fields-asterisk

amesplant committed:
Update how required field asterisk are displayed

amesplant committed:
Add back in the validate-required class

amesplant-dmv committed:
Add aria-required and update required asterisk

amesplant-dmv committed:
Add changelog

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

dkotter commented on plugins/woocommerce/includes/wc-template-functions.php:
```diff
@@ -2944,7 +2949,7 @@ function woocommerce_form_field( $key, $args, $value = null ) {
 					}
 
-					$field .= '</select>';
+					$field .= esc_attr( $asterisk ) . '</select>';
```
Since `$asterisk` isn't an attribute but is full HTML, `esc_attr` isn't the proper escaping function here. We may be able to leave this since we're not actually echoing anything here, just storing that in the `$field` variable where I assume it get's echoed later

dkotter commented on plugins/woocommerce/includes/wc-template-functions.php:
```diff
@@ -3054,7 +3059,7 @@ function woocommerce_form_field( $key, $args, $value = null ) {
 					foreach ( $args['options'] as $option_key => $option_text ) {
 						$field .= '<input type="radio" class="input-radio ' . esc_attr( implode( ' ', $args['input_class'] ) ) . '" value="' . esc_attr( $option_key ) . '" name="' . esc_attr( $key ) . '" ' . implode( ' ', $custom_attributes ) . ' id="' . esc_attr( $args['id'] ) . '_' . esc_attr( $option_key ) . '"' . checked( $value, $option_key, false ) . ' />';
-						$field .= '<label for="' . esc_attr( $args['id'] ) . '_' . esc_attr( $option_key ) . '" class="radio ' . implode( ' ', $args['label_class'] ) . '">' . esc_html( $option_text ) . '</label>';
+						$field .= '<label for="' . esc_attr( $args['id'] ) . '_' . esc_attr( $option_key ) . '" class="radio ' . implode( ' ', $args['label_class'] ) . '">' . esc_html( $option_text ) . esc_attr( $asterisk ) . '</label>';
```
I'd suggest removing the `esc_attr` here

dkotter reviewed (changes requested):
Left a few comments but also seeing a number of workflow failures, like linting and E2E. Those will need cleaned up before this can proceed.

amesplant committed:
Fix linting errors

amesplant committed:
Remove unnecessary esc_attr calls

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant committed:
Fix linting of spaces

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant committed:
Update tests for required radio and checkboxes

amesplant committed:
Update styles for the change from a link to a button for `.reset_varations`

amesplant committed:
Update test since the `.reset_variations` is now a button

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant-dmv committed:
Remove the asterisks from the test

amesplant-dmv committed:
Remove * since not longer used for required fields

amesplant-dmv committed:
Remove asterisks from tests

amesplant-dmv committed:
Remove (optional) on tests for fields as they have been removed from all fields

amesplant-dmv committed:
Fix linting error

amesplant-dmv committed:
Remove more asterisks

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant-dmv committed:
Update product varation test

amesplant-dmv committed:
Merge branch 'fix/48948-required-fields-asterisk' of https://github.com/amesplant/woocommerce into fix/48948-required-fields-asterisk

amesplant-dmv committed:
Remove additional fields with (optional) in their label

amesplant-dmv committed:
Add an extra space in Shipping address for testing puproses

amesplant-dmv committed:
Revert "Add an extra space in Shipping address for testing puproses"

This reverts commit 78dc51dddde4fe1dbee78a493cb6049004a9e03d.

amesplant-dmv committed:
Revert "Remove additional fields with (optional) in their label"

This reverts commit 2cd490be4e22abd0bd4aaef538aaea6a664bc6cd.

amesplant-dmv committed:
Add back in (optional) for non required fields

amesplant-dmv committed:
Update another field that should be tested as optional

amesplant-dmv committed:
Remove unused variable since it is immediately used below it

amesplant commented:
@dkotter I have finally have all failures working as expected

dkotter commented on plugins/woocommerce/templates/myaccount/form-edit-account.php:
```diff
@@ -12,35 +12,59 @@
+/**
+ * Hook - woocommerce_before_edit_account_form.
+ *
+ * @since 9.3.0
```
This hook wasn't added in 9.3.0. In doing some digging, it was introduced in this PR: https://github.com/woocommerce/woocommerce/pull/10975 which seems to be part of the 2.6.0 release
```suggestion
 * @since 2.6.0
```

amesplant committed:
Update plugins/woocommerce/templates/myaccount/form-edit-account.php

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant commented:
@dkotter I committed your version update

dkotter commented on plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout.spec.js:
```diff
@@ -175,10 +175,29 @@ test.describe(
 			await expect( page.locator( '#payment_method_cod' ) ).toBeEnabled();
 		} );
 
+		await test.step( 'Complete the checkout form', async () => {
```
Looks like E2E tests are failing because of the `await` here:

https://github.com/woocommerce/woocommerce/actions/runs/10373657102/job/28719326875?pr=49696

amesplant replied:
@dkotter hmmm .. I had all of these errors cleared up and looks like when `trunk` was merged in that we have a few more .. I will look into them and straighten them out.

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant committed:
Remove test for "Complete the checkout form"

amesplant committed:
Fix spacing issues

amesplant committed:
Fix additional settings

amesplant committed:
Remove 'required' asterisks from tests

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant commented:
@dkotter that error is fixed and ready for another review before we move it over to QA

dkotter reviewed (approved):
@ankitguptaindia @qasumitbagthariya This is ready for QA

amesplant committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Add aria-required attribute to email field on edit account page

Manussakis committed:
Fix how Phone and Email address fields are selected in checkout tests

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Add asterisk to address required fields

Manussakis committed:
Edit asterisk style to match asterisk character

Manussakis committed:
Bump form-edit-account template version

Manussakis commented:
@dkotter I have updated this PR a bit, please review it again:
1. Fixed the issue with the email field flagged by @qasumitbagthariya on [this comment](https://github.com/woocommerce/woocommerce/pull/49696#issuecomment-2293922720).
Commit: [Add aria-required attribute to email field on edit account page](https://github.com/woocommerce/woocommerce/pull/49696/commits/bf9fee34674197bf6db5cb130c7dfad48eed68b8)
2. Updated the tests: 
Commit: [Fix how Phone and Email address fields are selected in checkout tests](https://github.com/woocommerce/woocommerce/pull/49696/commits/59b3276097201541c91168938156336d42fd95cb)
3. Implemented the new approach to render the asterisk sign in the address fields on checkout.
Commit: [Add asterisk to address required fields](https://github.com/woocommerce/woocommerce/pull/49696/commits/9958020ac3689c529d67e63241e80303f43751a2)
4. Edited the asterisk style to match the asterisk character.
Commit: [Edit asterisk style to match asterisk character](https://github.com/woocommerce/woocommerce/pull/49696/commits/ce4439afb0e05129792e943f385421ad0a34964c)

Manussakis commented:
@qasumitbagthariya This PR is ready for another round of QA. Thanks!

qasumitbagthariya commented:
@Manussakis 

QA Report ⚠️
---

I've noticed an issue in this PR where the validation asterisk (*) is not visible in the following themes:

- Storefront
- Twenty Twenty-Three (TT3)

**Storefront**

![image](https://github.com/user-attachments/assets/25028a31-31cb-4ff3-8730-1c4adccbf8e0)


**TT3**

![image](https://github.com/user-attachments/assets/1079b8bb-8e1f-432a-9e81-32dd746f6eaa)

**TT2**
![image](https://github.com/user-attachments/assets/3cf38dcc-98a0-48ac-8d3f-169403f931ce)


**TT4** It seems to be working properly in Twenty Twenty-Four (TT4).

![screencapture-woo-core-local-my-account-edit-account-2024-09-04-15_03_49](https://github.com/user-attachments/assets/8eeabfe4-7235-4d0d-8819-85d1db8ce160)

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Create asterisk mixin

Manussakis committed:
Add asterisk style to T17 theme

Manussakis committed:
Add asterisk style to T19 theme

Manussakis committed:
Add asterisk style to TT theme

Manussakis committed:
Add asterisk style to TT1 theme

Manussakis committed:
Add asterisk style to TT2 theme

Manussakis committed:
Add asterisk style to TT3 theme

Manussakis committed:
Account for two dots elements to add aria-label to page link

Manussakis commented:
Thanks, @qasumitbagthariya for testing this PR!

I have fixed the issue you flagged. Once this PR passes code review, please test it again against the following themes:
- Twenty Seventeen
- Twenty Nineteen
- Twenty Twenty
- Twenty Twenty-One
- Twenty Twenty-Two
- Twenty Twenty-Three
- Twenty Twenty-Four

Regarding the StoreFront theme, it doesn't load the WooCommerce stylesheet so the asterisk style is not applied. This is likely a change that must be handled on the theme side. Let's flag that once we move this task to Woo review.

@dkotter Please re-review it again. Thanks!
- [Create asterisk mixin](https://github.com/woocommerce/woocommerce/pull/49696/commits/e308e9b0dbccf5f0fdbfe2dda042b4314dda987c)
- [Add asterisk style to T17 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/c82f8464662ea38865e04db990689093c098fd6d)
- [Add asterisk style to T19 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/45dd57e7aac570395cbbe1503b68a12c09e235f1)
- [Add asterisk style to TT theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/3ee5503e6a3cfe6a7af6c80a21dda90c004a29e7)
- [Add asterisk style to TT1 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/ab3ef333062ebdbb6a13f1d291f3bbd67bfa2f26)
- [Add asterisk style to TT2 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/6dd3fded22094e2f9aa804acdf322332caa02f25)
- [Add asterisk style to TT3 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/047b5358e37aa822ab87638ab15e37f7dea871e6)
- [Account for two dots elements to add aria-label to page link](https://github.com/woocommerce/woocommerce/pull/49696/commits/baebd51420bb714054a42cc392c12ebb230a5beb)

qasumitbagthariya commented:
> Thanks, @qasumitbagthariya for testing this PR!
> 
> I have fixed the issue you flagged. Once this PR passes code review, please test it again against the following themes:
> 
> * Twenty Seventeen
> * Twenty Nineteen
> * Twenty Twenty
> * Twenty Twenty-One
> * Twenty Twenty-Two
> * Twenty Twenty-Three
> * Twenty Twenty-Four
> 
> Regarding the StoreFront theme, it doesn't load the WooCommerce stylesheet so the asterisk style is not applied. This is likely a change that must be handled on the theme side. Let's flag that once we move this task to Woo review.
> 
> @dkotter Please re-review it again. Thanks!
> 
> * [Create asterisk mixin](https://github.com/woocommerce/woocommerce/pull/49696/commits/e308e9b0dbccf5f0fdbfe2dda042b4314dda987c)
> * [Add asterisk style to T17 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/c82f8464662ea38865e04db990689093c098fd6d)
> * [Add asterisk style to T19 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/45dd57e7aac570395cbbe1503b68a12c09e235f1)
> * [Add asterisk style to TT theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/3ee5503e6a3cfe6a7af6c80a21dda90c004a29e7)
> * [Add asterisk style to TT1 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/ab3ef333062ebdbb6a13f1d291f3bbd67bfa2f26)
> * [Add asterisk style to TT2 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/6dd3fded22094e2f9aa804acdf322332caa02f25)
> * [Add asterisk style to TT3 theme](https://github.com/woocommerce/woocommerce/pull/49696/commits/047b5358e37aa822ab87638ab15e37f7dea871e6)
> * [Account for two dots elements to add aria-label to page link](https://github.com/woocommerce/woocommerce/pull/49696/commits/baebd51420bb714054a42cc392c12ebb230a5beb)

@Manussakis 
**Quetion:**

Has the asterisk sign been removed from the email ID field in the latest fix? I can't see it in TT3, TT3, TT4. Could you please confirm?

Manussakis commented:
> Has the asterisk sign been removed from the email ID field in the latest fix? I can't see it in TT3, TT3, TT4. Could you please confirm?

Hey @qasumitbagthariya !
I am moving this task to in progress again and I will take a look once I get back to the project.

Thanks!

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis commented:
Hey @qasumitbagthariya 

> Has the asterisk sign been removed from the email ID field in the latest fix? I can't see it in TT3, TT3, TT4. Could you please confirm?

I checked the email field and I can see the asterisk sign on My Account page > Addresses > Edit Billing address.
Could you please confirm I am checking the right form?

- TT2
![image](https://github.com/user-attachments/assets/929d9fb6-0646-412b-b973-e82672779076)

- TT3
![image](https://github.com/user-attachments/assets/5f196870-6280-4f89-9e15-73a8cbf4f026)

- TT4
![image](https://github.com/user-attachments/assets/5035cdca-b1a3-4825-8c65-005dd1fef464)

qasumitbagthariya commented:
QA Update: ✅ 
----

It’s working for me now. I'm not sure why it wasn’t working before, but everything seems to be functioning as expected now


**TT3**

<img width="1401" alt="image" src="https://github.com/user-attachments/assets/7eaa2039-cab9-4996-9fbe-8e671ce45cc1">

**TT4**

<img width="1365" alt="image" src="https://github.com/user-attachments/assets/2c84ffc1-4632-44b7-884b-8c612d00b0df">

**TT2**

<img width="1344" alt="image" src="https://github.com/user-attachments/assets/4de3f8a5-9841-4b9d-a7e8-bf75fb56930d">


**Testing Environment** 

<Details>

- WordPress: 6.6
- WooCommerce - WooCommerce - Dev version with this fix specific branch.
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Git Branch: amesplant:fix/48948-required-fields-asterisk

</Details>

**Status update:**
@vikrampm1 Ready for the next steps.

vikrampm1 commented:
@Manussakis please remove [Draft] from the title and make the PR ready for review.

Manussakis commented:
@vikrampm1 For some reason I don't have access to update this PR.
@amesplant Would you mind making it ready for review and removing [Draft] from the title?

vikrampm1 commented:
@mikejolley / @woocommerce/rubik this PR is now ready for your review!

alexflorisca commented:
> @mikejolley / @woocommerce/rubik this PR is now ready for your review!
Thanks for the ping. Can I kindly ask that next time you ping only the person assigned to review the PR? 🙏

mikejolley reviewed (commented):
Thanks for preparing this PR. I have a few questions before reviewing fully I hope you can clarify.

1. Pure CSS asterisk.

Can you explain why this is done using span elements and CSS vs `content` property? I am just concerned a pure-css asterisk would be more prone to breakage by 3rd party themes. Would it be simpler (and use less markup) to use:

```
.asterisk::after { content: "*"; }
```

2. `aria-hidden` support

With the original solution and `aria-hidden`, was this not suitable because some screen readers ignore it? Are other properties like `speak` supported? We've used `aria-hidden` in other places so I wonder how bad the experience is or if its a major issue. 

3. Classname change

If point 2 means the * has to be removed from the markup, can we keep the same `.required` markup so some styles are preserved?

Thanks in advance.

Manussakis commented:
@mikejolley 

> 1. Pure CSS asterisk.
> 2. aria-hidden support
> 3. Classname change

I believe all your questions are discussed in detail in issue #48948.
Would you mind checking the thread and let me know if you have any additional questions?

Thanks!

mikejolley commented:
@Manussakis 

> Pure CSS asterisk.

The other post says "Adding the asterisk, using a bit of html and css (which are empty and thus will not communicate anything to a screen reader)". 

I would like clarification to why `content` prop would be worse in this case, and which screen readers, if any, would ignore all directives and announce the text. I believe using a character via `content` prop would be less prone to theme conflicts.

> aria-hidden support

The post says "That way when users using a screen reader have focus on the input field, the label is announced with the additional "required" text, without relying on aria-hidden (which is not respected by all screen readers, ahem NVDA)"

But does not mention `speak` nor whether this applies to content injected via CSS rather than in the DOM.

> Classname change

Not addressed in the other post.

vikrampm1 commented:
Moving this back to In-Progress until @Manussakis is back from PTO and able to review the above.

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

amesplant commented:
@mikejolley
I can definitely understand your pause for concern, and hopefully I can shed some more light onto why some of these decisions were made.

> 1. Pure CSS asterisk.
> 
> Can you explain why this is done using span elements and CSS vs `content` property? I am just concerned a pure-css asterisk would be more prone to breakage by 3rd party themes. Would it be simpler (and use less markup) to use:
> 
> ```
> .asterisk::after { content: "*"; }
> ```
> 
I would concur that this is the path that many sites take, but due to the inconsistencies in how assistive technologies interpret pseudo elements, they may be either completely ignored or misunderstood, causing confusion for users. Some will read the * as "asterisk", while others will completely ignore it. 

> 2. `aria-hidden` support
> 
> With the original solution and `aria-hidden`, was this not suitable because some screen readers ignore it? Are other properties like `speak` supported? We've used `aria-hidden` in other places so I wonder how bad the experience is or if its a major issue.
> 
Again, another method that is often used by sites, and while widely supported, `aria-hidden` is not always respected consistently by all screen readers, such as NVDA, depending on the context and how the element is structured in the DOM.

Screen readers do not typically honor the speak property, as they rely on semantic HTML and ARIA attributes rather than CSS for determining what to announce. 


> 3. Classname change
> 
> If point 2 means the * has to be removed from the markup, can we keep the same `.required` markup so some styles are preserved?
> 
> Thanks in advance.
I believe that we have not removed the `.required` class, and instead replaced with `.asterisk` where appropriate.


I hope this addresses your concerns and provides a better picture for the decisions we made to make this more accessible.

mikejolley commented:
@amesplant thanks. I've asked the team to take a look at this while I am AFK. The asterisk is the thing I am most unsure about; struggling to find examples to back this one up. I cannot believe we'd be the only project to run into this.

nielslange commented:
> I believe that we have not removed the .required class, and instead replaced with .asterisk where appropriate.

@amesplant I'm just going through the discussion that took place in this PR and on #48948. As for the CSS class name change, which @mikejolley mentioned in https://github.com/woocommerce/woocommerce/pull/49696#pullrequestreview-2381459329, it appears that the `.required` class name has indeed been removed. Before this PR, I could see the following markup:

```html
<label for="account_first_name">
	"First name&nbsp;"
	<span class="required">*</span>
</label>
```

With this PR, I see the following markup:

```html
<label for="account_first_name">
	" First name "
	<span class="asterisk">
		<span class="asterisk-line"></span>
		<span class="asterisk-line"></span>
		<span class="asterisk-line"></span>
	</span>
</label>
```

This will affect merchants that overwrite the `.require` CSS class name. For backwards-compatibility, I propose that we keep it. 

I also noticed that for the billing and shipping address forms, the space between the label and the asterisk disappeared. The section, that's causing this, can be found on https://github.com/woocommerce/woocommerce/blob/9b925c79408aa5f9fd78120cd38242320d8b81b0/plugins/woocommerce/includes/wc-template-functions.php#L2945-L2949

I propose adding a `&nbsp;` before the `<span>`, thus changing 

```php
$required_indicator = '<span class="asterisk">[...]'; 
```

to

```php
$required_indicator = '&nbsp;<span class="asterisk">[...]'; 
```

As for the other question, I'll need to review your PR more in depth. That said, I wanted to share my findings already with you.

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Add required class and white space back to required fields

Manussakis committed:
Add missing required class

Manussakis commented:
@nielslange I will work on the fixes you flagged in [this comment](https://github.com/woocommerce/woocommerce/pull/49696#issuecomment-2500163342) and let you know when they're done.

Thanks!

Manussakis committed:
Fix radio required field test

Manussakis commented:
@nielslange I have fixed [these issues](https://github.com/woocommerce/woocommerce/pull/49696#issuecomment-2500163342).

nielslange commented:
@amberhinds I've reviewed the PR, and the changes function as expected. However, I'm not entirely convinced that the proposed solution is the best approach. As @qasumitbagthariya highlighted in [this comment](https://github.com/woocommerce/woocommerce/pull/49696#issuecomment-2328389841), the asterisks are no longer visible when using the Storefront theme. Given that Storefront remains one of the most popular themes for WooCommerce, I believe we should aim for a solution that is backward-compatible and doesn’t require themes to adapt to our implementation.

Regarding the current approach of building the asterisk using CSS, I couldn’t find resources advocating for this method. Could you share references or examples that explain the rationale behind this solution? As @mikejolley suggested in [his review](https://github.com/woocommerce/woocommerce/pull/49696#pullrequestreview-2381459329), I believe using the aria-hidden attribute might be a better direction. While I understand your concern about screen readers inconsistently respecting aria-hidden, I feel we should carefully weigh the trade-offs of this workaround against the broader implications for accessibility standards.

Looking at [this example](https://equalizedigital.com/contact/), I wonder if we might consider an alternative approach. Instead of implementing asterisks, what if we displayed "(Required)" after the corresponding form fields? This might offer a clearer and more universally accessible solution.

---

@elizaan36 What are your thoughts on using "(Required)" instead of "*" to indicate required form fields within the My Account section? Here's an example of what I mean:

<table>
<tr>
<td valign="top">Asterisk:
<br><br>
<img width="624" alt="Screenshot 2024-12-03 at 16 56 54" src="https://github.com/user-attachments/assets/758e348f-5453-4e06-ba30-bd2cbcb5629b">
</td>
<td valign="top">(Required):
<br><br>
<img width="622" alt="Screenshot 2024-12-03 at 16 58 57" src="https://github.com/user-attachments/assets/fa6cc2d8-b741-4708-ba3a-151aa8b0075d">
</td>
</tr>
</table>

elizaan36 commented:
hey, Thanks so much for working on this to improve accessibility at checkout. 

@nielslange No, I don't think the full word should be written out. The asterisk is sufficient to indicate a required field, and should be accompanied by immediate validation of form field entries. The word "Required" on each field would add unnecessary noise to the checkout experience.

nielslange commented:
> @nielslange No, I don't think the full word should be written out. The asterisk is sufficient to indicate a required field, and should be accompanied by immediate validation of form field entries. The word "Required" on each field would add unnecessary noise to the checkout experience.

Thank you for your response, @elizaan36. In that case, we’ll proceed with using the asterisk.

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

nielslange commented:
@amesplant I wonder if you have seen my question in https://github.com/woocommerce/woocommerce/pull/49696#issuecomment-2514078874.

amberhinds commented:
Just chiming to say that I agree that it would be better to add the asterisk as text that is hidden via aria-hidden rather than using CSS to add it. That's what is more conventionally done. I'd worry about the asterisk not appearing if CSS is missing for one reason or another.

amesplant commented:
I also agree with the issue of removing the asterisk based on it being used in other themes and not allowing it to be backwards compatible.

nielslange commented:
> I also agree with the issue of removing the asterisk based on it being used in other themes and not allowing it to be backwards compatible.

@amesplant Can you take care of implementing @amberhinds' [suggestion](https://github.com/woocommerce/woocommerce/pull/49696#issuecomment-2540316578) and let me know when I can review the PR again?

Manussakis commented:
@nielslange I will take care of implementing the @amberhinds 's suggestions 😉 

cc: @amesplant

nielslange commented:
> @nielslange I will take care of implementing the @amberhinds 's suggestions 😉

Thanks, @Manussakis. 🙌

nielslange reviewed (commented):
Hello @Manussakis, could you let me know when this PR is ready for review? I'm submitting the review form for now, as I get daily reminders.

Manussakis commented:
>  could you let me know when this PR is ready for review?

@nielslange I plan to work on it today. I will ping you once it's done.

Thanks!

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Remove asterisk styles

Manussakis committed:
Remove asterisk HTML markup

Manussakis committed:
Remove asterist character from tests

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis committed:
Add (optional) text to Company label in test

Manussakis committed:
Revert change in Company toggle on the editor

Manussakis commented:
@dkotter I have updated this PR to implement @amberhinds 's suggestions. It's ready for another round of code review now.
@qasumitbagthariya This PR will need another round of QA once the code review is done.
@nielslange Once the reviews on the 10up side have been completed, I will send it back to you.

Thank you all!

Manussakis commented:
@dkotter I have updated this PR to implement @amberhinds 's suggestions. It's ready for another round of code review now.
@qasumitbagthariya This PR will need another round of QA once the code review is done.
@nielslange Once the reviews on the 10up side have been completed, I will send it back to you.

**HEADS-UP**
There is a [test that failed](https://github.com/woocommerce/woocommerce/actions/runs/12400819513/job/34619073150#step:9:267) but it doesn't seem to be related to the changes in this PR.
The Company switch in the Shipping Address block is not changing its state during the tests.
![Screenshot 2024-12-18 at 18 51 43](https://github.com/user-attachments/assets/bd04210f-dbd1-47f0-a065-7b2db4f0dbf2)

Manussakis committed:
Remove duplicate condition to check required input

Manussakis committed:
Align variables

qasumitbagthariya commented:
### QA/Test Report ✅ 
-------------

I have checked this PR in the `amesplant:fix/48948-required-fields-asterisk` branch and changes working as expected.

Verify with below themes.

- TT4
- TT5
- Storefront

**aria-hidden**

<img width="1192" alt="image" src="https://github.com/user-attachments/assets/60bd84ac-e5a7-458b-be17-3b4ad4aa6530" />


https://github.com/user-attachments/assets/f30e9546-0f40-41c6-a2ea-af2dc0963e19


**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Storefront 4.6.0
- WooCommerce - WooCommerce - Dev version with this fix specific branch.
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2
- Git Branch: amesplant:fix/48948-required-fields-asterisk

</Details>

cc @vikrampm1 
Ready for the next step.

jeffpaul commented:
@nielslange this is ready for your testing/review now, thanks!

nielslange commented:
> @nielslange this is ready for your testing/review now, thanks!

Thanks for your ping, @jeffpaul. I'm reviewing this PR now.

nielslange commented on plugins/woocommerce/templates/myaccount/form-edit-account.php:
```diff
@@ -17,31 +17,38 @@
 
```
The version of this template needs to be bumped to `9.7.0`.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -749,11 +749,11 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
 			// Check text inputs in edit mode match the expected values.
 			const govIdInput = checkoutPageObject.page.getByLabel(
-				'Government ID *',
+				'Government ID',
```
```suggestion
				'Government ID *',
```

Due to `{ exact: true }`, removing the asterisk causes this e2e test to fail.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -805,11 +805,11 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
 			// Check text inputs in edit mode match the expected values.
 			const shippingGovIdInput = checkoutPageObject.page.getByLabel(
-				'Government ID *',
+				'Government ID',
```
```suggestion
				'Government ID *',
```

Due to `{ exact: true }`, removing the asterisk causes this e2e test to fail.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -749,11 +749,11 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
 			);
 			const confirmGovIdInput = checkoutPageObject.page.getByLabel(
-				'Confirm government ID *',
+				'Confirm government ID',
```
```suggestion
				'Confirm government ID *',
```

Due to `{ exact: true }`, removing the asterisk causes this e2e test to fail.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -805,11 +805,11 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
 			);
 			const shippingConfirmGovIdInput =
-				checkoutPageObject.page.getByLabel( 'Confirm government ID *', {
+				checkoutPageObject.page.getByLabel( 'Confirm government ID', {
```
```suggestion
				'Confirm government ID *',
```

Due to `{ exact: true }`, removing the asterisk causes this e2e test to fail.

nielslange reviewed (commented):
Thanks for working on this, @amesplant and @Manussakis. The PR works as expected, but the CI shows two errors:
- The template version needs to be bumped.
- One e2e tests due to its label definition.

Manussakis committed:
Add asterisk to locator to fix e2e test

Co-authored-by: Niels Lange <niels.lange@a8c.com>

Manussakis committed:
Add asterisk to locator to fix e2e test

Co-authored-by: Niels Lange <niels.lange@a8c.com>

Manussakis committed:
Add asterisk to locator to fix e2e test

Co-authored-by: Niels Lange <niels.lange@a8c.com>

Manussakis committed:
Add asterisk to locator to fix e2e test

Manussakis committed:
Update single-product-review template version

Manussakis committed:
Merge branch 'trunk' into fix/48948-required-fields-asterisk

Manussakis commented:
@nielslange I have fixed the issues you flagged. Please re-review when you can 😉

nielslange reviewed (approved):
> @nielslange I have fixed the issues you flagged. Please re-review when you can 😉

Thanks for addressing the open points, @Manussakis. I've successfully tested this PR and all CI tests passed as well. I'll merge this PR now. Thanks again for working on this. 🙌

PaulSchiretz commented:
Hi @nielslange, this this pull request produces invalid html as select tags cannot contain span tags which are added to the country select fields for the checkout after this pull request. which triggers a console warning in chrome and doesn't validate with any w3c validator.

Here are the 2 lines that cause the issue:
[https://github.com/woocommerce/woocommerce/commit/c661b9f1edd0f2b46b65056359d33724878e3679#diff-d49488c34017d273db858b7b6eff58f22d95310991d26638e75793c6a1e98678R3084](https://github.com/woocommerce/woocommerce/commit/c661b9f1edd0f2b46b65056359d33724878e3679#diff-d49488c34017d273db858b7b6eff58f22d95310991d26638e75793c6a1e98678R3084)

[https://github.com/woocommerce/woocommerce/commit/c661b9f1edd0f2b46b65056359d33724878e3679#diff-d49488c34017d273db858b7b6eff58f22d95310991d26638e75793c6a1e98678R3144](https://github.com/woocommerce/woocommerce/commit/c661b9f1edd0f2b46b65056359d33724878e3679#diff-d49488c34017d273db858b7b6eff58f22d95310991d26638e75793c6a1e98678R3144)

Can you have a look at this?

nielslange commented:
Thanks for reaching out, @PaulSchiretz. I'm afraid this is a tricky issue, as you can tell from the number of comments. We've looked into various options and considered the current approach to be the least error-prone while making WooCommerce more accessible. That said, I agree that it's not ideal that the implementation results in invalid HTML markup.  

We'll keep #55947 open to look into this in the future, but unfortunately, I don’t have time to work on it at the moment.

ralucaStan commented:
This PR caused a duplicated `*` symbol for the Checkout Fields Editor plugin. It will be solved in that plugin.

https://github.com/user-attachments/assets/02a39652-d945-41b1-a8b2-b75760965f08
